module3/homework/advanced_calculator/mathOperation.cpp

```diff
@@ -0,0 +1,158 @@
+#include "mathOperation.hpp"
+#include "errorCodeEnum.hpp"
+#include <cmath>
+#include <cctype>
+#include <iostream>
+#include <vector>
+
+mathOperation::mathOperation(std::string& inputString)
+    : dataString_(inputString), result_(0), firstValue_(0), secondValue_(0),
+      operationCharacter_(""), actualError_(ErrorCode::OK) {
+        removeSpaces();
+        actualError_ =  checkIfBadFormat();
+        actualError_ =  checkIfBadCharacter();
+        actualError_ =  checkIfDivideByZero();
+        actualError_ = checkIfSqrtOfNegativeNumber();
+        actualError_ = checkIfmoduleOfNonIntegerValue();
+        if(actualError_ == ErrorCode::OK) calculateResult();
+
+      };
+
+double mathOperation::getResult() const { return result_; }
+ErrorCode mathOperation::getErrorCode() const { return actualError_; }
+
+
+double mathOperation::add(double &firstValue, double &secondValue) {
+  return firstValue + secondValue;
+}
+double mathOperation::subtract(double &firstValue, double &secondValue) {
+  return firstValue - secondValue;
+}
+double mathOperation::multiplicate(double &firstValue, double &secondValue) {
+  return firstValue * secondValue;
+}
+double mathOperation::divide(double &firstValue, double &secondValue) {
+  return firstValue / secondValue;
+}
+double mathOperation::modulo(double &firstValue, double &secondValue) {
+  return static_cast<double>(static_cast<int>(firstValue)%
+                             static_cast<int>(secondValue));
+}
+double mathOperation::factorial(double &firstValue, double &secondValue) {
+  if(firstValue < 0){
+    return 1;
+  }else return static_cast<double>(std::tgamma(firstValue+1));
```

Comment:
```suggestion
  return firstValue < 0 ? 1.0 : static_cast<double>(std::tgamma(firstValue+1));
```
You can consider simplifying it by lambda expression.